// openclaw@2026.3.8/dist/program-BY--_6zJ.js
import { C as setVerbose, H as getCommandPathWithRootOptions, J as hasFlag, L as ALLOWED_LOG_LEVELS, O as isRich, X as hasRootVersionAlias, Y as hasHelpOrVersion, f as defaultRuntime, k as theme, q as getVerboseFlag, z as tryParseLogLevel } from "./subsystem-kzdGVyce.js";
import "./paths-BfR2LXbA.js";
import "./boolean-BHdNsbzF.js";
import { r as isTruthyEnvValue } from "./entry.js";
import "./auth-profiles-UpqQjKB-.js";
import { n as replaceCliName, r as resolveCliName } from "./command-format-CO3N-oGG.js";
import "./agent-scope-BDXYqF89.js";
import { c as escapeRegExp } from "./utils-B4wShrgI.js";
import "./boundary-file-read-CcBHg_z-.js";
import "./logger-BWbD2ty6.js";
import "./exec-CDUUaRm4.js";
import "./github-copilot-token-CcBrBN3h.js";
import "./host-env-security-blJbxyQo.js";
import { t as VERSION } from "./version-Bxx5bg6l.js";
import "./registry-Gs6xb3DK.js";
import "./manifest-registry-CFnC0yb4.js";
import "./dock-6TSCZqEs.js";
import "./models-config-Ct5bZG7t.js";
import "./tokens-B2BIDASk.js";
import "./compact-D3emcZgv.js";
import "./plugins-v0UiRqAc.js";
import "./accounts-CFByy36g.js";
import "./channel-config-helpers-Cc37-fbu.js";
import "./send-CbPREZmS.js";
import "./send-Bj6Jq5a4.js";
import "./with-timeout-BbQnczA2.js";
import "./deliver-bv9qu6sz.js";
import "./diagnostic-BjGEEiT0.js";
import "./accounts-Dd357WkV.js";
import "./image-ops-BZX4S9OA.js";
import "./send-oxP8y9QI.js";
import "./pi-model-discovery-CHZFD2LJ.js";
import "./message-channel-FOsBizRL.js";
import "./pi-embedded-helpers-nnm4iOi3.js";
import "./sandbox-D5S9pIUQ.js";
import "./tool-catalog-DAnzECvM.js";
import "./chrome-_dzns-pF.js";
import "./tailscale-DwUR_aBY.js";
import "./tailnet-BEiNr4qb.js";
import "./ws-BzHIlqUk.js";
import "./auth-ztX1XT7i.js";
import "./credentials-UUOBfSK0.js";
import "./resolve-configured-secret-input-string-CALUVGLW.js";
import "./server-context-Cysq8uwv.js";
import "./frontmatter-B5Xx_5Cp.js";
import "./env-overrides-CghEbWu_.js";
import "./path-alias-guards-2B2VwR9Z.js";
import "./skills-BcTP9HTD.js";
import "./paths-Bv0Fd5hm.js";
import "./proxy-env-DBXIh4R2.js";
import "./redact-b9XS0Muh.js";
import "./errors-oTqWODa8.js";
import "./fs-safe-DnLkCsXk.js";
import "./store-Cpd6HHUc.js";
import "./ports-DwmubRiU.js";
import "./trash-CKkKoeyk.js";
import "./server-middleware-D5JzODMR.js";
import "./sessions-Bh025Q5O.js";
import "./paths-D3cXuvcv.js";
import "./chat-envelope-Ba8wEdTK.js";
import "./tool-images-CpDLyo5v.js";
import "./thinking-CHrWXrB4.js";
import "./exec-approvals-allowlist-CmGNghDQ.js";
import "./exec-safe-bin-runtime-policy-BIU2HUmH.js";
import "./model-catalog-RhG2KE3w.js";
import "./fetch-CEXnYKAv.js";
import "./audio-transcription-runner-CkHnvIGC.js";
import "./fetch-guard-W2TcXMQH.js";
import "./image-CiVJHtoW.js";
import "./tool-display-BJIcX_xB.js";
import "./api-key-rotation-BQqyYKRt.js";
import "./proxy-fetch-BJuVlAGN.js";
import "./ir-MPXYeM9z.js";
import "./render-DT-umBSz.js";
import "./target-errors-DkGMVAsH.js";
import "./commands-7gthnTya.js";
import "./commands-registry-Cw8jiBIe.js";
import "./session-cost-usage-0f4Id4RU.js";
import "./session-utils-Cff6Vxd-.js";
import { n as resolveCommitHash } from "./git-commit-Z4N0ntjy.js";
import "./sqlite-Q5Gkerf1.js";
import "./call-BfhGytph.js";
import "./pairing-token-Bu2Xt1ri.js";
import "./fetch-Q5enmqDF.js";
import "./pairing-store-v0j3oBoa.js";
import "./exec-approvals-Bc3G-ceB.js";
import "./nodes-screen-DPo01McV.js";
import "./restart-8rBxGr-n.js";
import "./system-run-command-DKHL4lN5.js";
import "./skill-commands-eL6X2qb6.js";
import "./pi-tools.policy-B0vneueD.js";
import "./workspace-dirs-Bi16CKb9.js";
import "./channel-activity-CoKobep6.js";
import "./tables--wgAB_S9.js";
import "./runtime-lifecycle-CI5BW-bB.js";
import "./stagger-BBJI9ikk.js";
import "./channel-selection-CZXcbBEr.js";
import "./plugin-auto-enable-Djxxngde.js";
import "./send-DIPIlK8e.js";
import "./outbound-attachment-DfwwWWQf.js";
import "./delivery-queue-cVcGAG6r.js";
import "./send-DHcDLbo7.js";
import "./proxy-3O29IgJK.js";
import "./timeouts-r52PcxXY.js";
import "./runtime-config-collectors-mVzg9wUg.js";
import "./command-secret-targets-tuZ5jpnm.js";
import "./connection-auth-wdU52rBJ.js";
import "./onboard-helpers-Be9c-Ktq.js";
import "./prompt-style-DMLxb9pN.js";
import "./pairing-labels-D1eHodye.js";
import "./memory-cli-C0jP3BCc.js";
import "./manager-_AfSnopa.js";
import { t as formatDocsLink } from "./links-CVgNb7TN.js";
import "./cli-utils-ChNkua_i.js";
import "./help-format-UvCfCIbo.js";
import "./progress-BVJ8qDnh.js";
import "./plugin-registry-By6DKgNN.js";
import { n as resolveCliChannelOptions } from "./channel-options-Ba-H44Hh.js";
import { t as getSubCliCommandsWithSubcommands } from "./register.subclis-B3aOfT8c.js";
import { a as registerProgramCommands, r as getCoreCliCommandsWithSubcommands } from "./command-registry--bBQl6ys.js";
import { r as setProgramContext } from "./program-context-Crxw5U5X.js";
import "./ports-CRpuvPaX.js";
import { n as formatCliBannerLine, r as hasEmittedCliBanner, t as emitCliBanner } from "./banner-CB1kDc3O.js";
import { Command, InvalidArgumentError } from "commander";
//#region src/cli/program/context.ts
function createProgramContext() {
	let cachedChannelOptions;
	const getChannelOptions = () => {
		if (cachedChannelOptions === void 0) cachedChannelOptions = resolveCliChannelOptions();
		return cachedChannelOptions;
	};
	return {
		programVersion: VERSION,
		get channelOptions() {
			return getChannelOptions();
		},
		get messageChannelOptions() {
			return getChannelOptions().join("|");
		},
		get agentChannelOptions() {
			return ["last", ...getChannelOptions()].join("|");
		}
	};
}
//#endregion
//#region src/cli/log-level-option.ts
const CLI_LOG_LEVEL_VALUES = ALLOWED_LOG_LEVELS.join("|");
function parseCliLogLevelOption(value) {
	const parsed = tryParseLogLevel(value);
	if (!parsed) throw new InvalidArgumentError(`Invalid --log-level (use ${CLI_LOG_LEVEL_VALUES})`);
	return parsed;
}
//#endregion
//#region src/cli/program/help.ts
const CLI_NAME = resolveCliName();
const CLI_NAME_PATTERN = escapeRegExp(CLI_NAME);
const ROOT_COMMANDS_WITH_SUBCOMMANDS = new Set([...getCoreCliCommandsWithSubcommands(), ...getSubCliCommandsWithSubcommands()]);
const ROOT_COMMANDS_HINT = "Hint: commands suffixed with * have subcommands. Run <command> --help for details.";
const EXAMPLES = [
	["openclaw models --help", "Show detailed help for the models command."],
	["openclaw channels login --verbose", "Link personal WhatsApp Web and show QR + connection logs."],
	["openclaw message send --target +15555550123 --message \"Hi\" --json", "Send via your web session and print JSON result."],
	["openclaw gateway --port 18789", "Run the WebSocket Gateway locally."],
	["openclaw --dev gateway", "Run a dev Gateway (isolated state/config) on ws://127.0.0.1:19001."],
	["openclaw gateway --force", "Kill anything bound to the default gateway port, then start it."],
	["openclaw gateway ...", "Gateway control via WebSocket."],
	["openclaw agent --to +15555550123 --message \"Run summary\" --deliver", "Talk directly to the agent using the Gateway; optionally send the WhatsApp reply."],
	["openclaw message send --channel telegram --target @mychat --message \"Hi\"", "Send via your Telegram bot."]
];
function configureProgramHelp(program, ctx) {
	program.name(CLI_NAME).description("").version(ctx.programVersion).option("--dev", "Dev profile: isolate state under ~/.openclaw-dev, default gateway port 19001, and shift derived ports (browser/canvas)").option("--profile <name>", "Use a named profile (isolates OPENCLAW_STATE_DIR/OPENCLAW_CONFIG_PATH under ~/.openclaw-<name>)").option("--log-level <level>", `Global log level override for file + console (${CLI_LOG_LEVEL_VALUES})`, parseCliLogLevelOption);
	program.option("--no-color", "Disable ANSI colors", false);
	program.helpOption("-h, --help", "Display help for command");
	program.helpCommand("help [command]", "Display help for command");
	program.configureHelp({
		sortSubcommands: true,
		sortOptions: true,
		optionTerm: (option) => theme.option(option.flags),
		subcommandTerm: (cmd) => {
			const hasSubcommands = cmd.parent === program && ROOT_COMMANDS_WITH_SUBCOMMANDS.has(cmd.name());
			return theme.command(hasSubcommands ? `${cmd.name()} *` : cmd.name());
		}
	});
	const formatHelpOutput = (str) => {
		let output = str;
		if (new RegExp(`^Usage:\\s+${CLI_NAME_PATTERN}\\s+\\[options\\]\\s+\\[command\\]\\s*$`, "m").test(output) && /^Commands:/m.test(output)) output = output.replace(/^Commands:/m, `Commands:\n  ${theme.muted(ROOT_COMMANDS_HINT)}`);
		return output.replace(/^Usage:/gm, theme.heading("Usage:")).replace(/^Options:/gm, theme.heading("Options:")).replace(/^Commands:/gm, theme.heading("Commands:"));
	};
	program.configureOutput({
		writeOut: (str) => {
			process.stdout.write(formatHelpOutput(str));
		},
		writeErr: (str) => {
			process.stderr.write(formatHelpOutput(str));
		},
		outputError: (str, write) => write(theme.error(str))
	});
	if (hasFlag(process.argv, "-V") || hasFlag(process.argv, "--version") || hasRootVersionAlias(process.argv)) {
		const commit = resolveCommitHash({ moduleUrl: import.meta.url });
		console.log(commit ? `OpenClaw ${ctx.programVersion} (${commit})` : `OpenClaw ${ctx.programVersion}`);
		process.exit(0);
	}
	program.addHelpText("beforeAll", () => {
		if (hasEmittedCliBanner()) return "";
		const rich = isRich();
		return `\n${formatCliBannerLine(ctx.programVersion, { richTty: rich })}\n`;
	});
	const fmtExamples = EXAMPLES.map(([cmd, desc]) => `  ${theme.command(replaceCliName(cmd, CLI_NAME))}\n    ${theme.muted(desc)}`).join("\n");
	program.addHelpText("afterAll", ({ command }) => {
		if (command !== program) return "";
		const docs = formatDocsLink("/cli", "docs.openclaw.ai/cli");
		return `\n${theme.heading("Examples:")}\n${fmtExamples}\n\n${theme.muted("Docs:")} ${docs}\n`;
	});
}
//#endregion
//#region src/cli/program/preaction.ts
function setProcessTitleForCommand(actionCommand) {
	let current = actionCommand;
	while (current.parent && current.parent.parent) current = current.parent;
	const name = current.name();
	const cliName = resolveCliName();
	if (!name || name === cliName) return;
	process.title = `${cliName}-${name}`;
}
const PLUGIN_REQUIRED_COMMANDS = new Set([
	"message",
	"channels",
	"directory",
	"agents",
	"configure",
	"onboard",
	"status",
	"health"
]);
const CONFIG_GUARD_BYPASS_COMMANDS = new Set([
	"backup",
	"doctor",
	"completion",
	"secrets"
]);
const JSON_PARSE_ONLY_COMMANDS = new Set(["config set"]);
let configGuardModulePromise;
let pluginRegistryModulePromise;
function shouldBypassConfigGuard(commandPath) {
	const [primary, secondary] = commandPath;
	if (!primary) return false;
	if (CONFIG_GUARD_BYPASS_COMMANDS.has(primary)) return true;
	if (primary === "config" && secondary === "validate") return true;
	return false;
}
function loadConfigGuardModule() {
	configGuardModulePromise ??= import("./config-guard-BX9AZYkh.js").then((n) => n.t);
	return configGuardModulePromise;
}
function loadPluginRegistryModule() {
	pluginRegistryModulePromise ??= import("./plugin-registry-By6DKgNN.js").then((n) => n.n);
	return pluginRegistryModulePromise;
}
function getRootCommand(command) {
	let current = command;
	while (current.parent) current = current.parent;
	return current;
}
function getCliLogLevel(actionCommand) {
	const root = getRootCommand(actionCommand);
	if (typeof root.getOptionValueSource !== "function") return;
	if (root.getOptionValueSource("logLevel") !== "cli") return;
	const logLevel = root.opts().logLevel;
	return typeof logLevel === "string" ? logLevel : void 0;
}
function isJsonOutputMode(commandPath, argv) {
	if (!hasFlag(argv, "--json")) return false;
	const key = `${commandPath[0] ?? ""} ${commandPath[1] ?? ""}`.trim();
	if (JSON_PARSE_ONLY_COMMANDS.has(key)) return false;
	return true;
}
function registerPreActionHooks(program, programVersion) {
	program.hook("preAction", async (_thisCommand, actionCommand) => {
		setProcessTitleForCommand(actionCommand);
		const argv = process.argv;
		if (hasHelpOrVersion(argv)) return;
		const commandPath = getCommandPathWithRootOptions(argv, 2);
		if (!(isTruthyEnvValue(process.env.OPENCLAW_HIDE_BANNER) || commandPath[0] === "update" || commandPath[0] === "completion" || commandPath[0] === "plugins" && commandPath[1] === "update")) emitCliBanner(programVersion);
		const verbose = getVerboseFlag(argv, { includeDebug: true });
		setVerbose(verbose);
		const cliLogLevel = getCliLogLevel(actionCommand);
		if (cliLogLevel) process.env.OPENCLAW_LOG_LEVEL = cliLogLevel;
		if (!verbose) process.env.NODE_NO_WARNINGS ??= "1";
		if (shouldBypassConfigGuard(commandPath)) return;
		const suppressDoctorStdout = isJsonOutputMode(commandPath, argv);
		const { ensureConfigReady } = await loadConfigGuardModule();
		await ensureConfigReady({
			runtime: defaultRuntime,
			commandPath,
			...suppressDoctorStdout ? { suppressDoctorStdout: true } : {}
		});
		if (PLUGIN_REQUIRED_COMMANDS.has(commandPath[0])) {
			const { ensurePluginRegistryLoaded } = await loadPluginRegistryModule();
			ensurePluginRegistryLoaded();
		}
	});
}
//#endregion
//#region src/cli/program/build-program.ts
function buildProgram() {
	const program = new Command();
	const ctx = createProgramContext();
	const argv = process.argv;
	setProgramContext(program, ctx);
	configureProgramHelp(program, ctx);
	registerPreActionHooks(program, ctx.programVersion);
	registerProgramCommands(program, ctx, argv);
	return program;
}
//#endregion
export { buildProgram };
