// openclaw@2026.3.8/dist/subagent-registry-runtime-mEjBIFwi.js
import "./paths-BJV7vkaX.js";
import "./subsystem-4K-e3L3i.js";
import "./utils-BhZbo8Nw.js";
import "./thinking-BYwvlJ3S.js";
import { b as resolveRequesterForChildSession, f as countActiveDescendantRuns, g as isSubagentSessionRunActive, m as countPendingDescendantRunsExcludingRun, p as countPendingDescendantRuns, v as listSubagentRunsForRequester, x as shouldIgnorePostCompletionAnnounceForSession, y as replaceSubagentRunAfterSteer } from "./reply-DeXK9BLT.js";
import "./agent-scope-CZF6h_5g.js";
import "./openclaw-root-DrFjwUcG.js";
import "./logger-BDhFOulu.js";
import "./exec-WYU5B_af.js";
import "./model-selection-Dovilo6b.js";
import "./github-copilot-token-D37fjdwy.js";
import "./boolean-CJxfhBkG.js";
import "./env-BL7t1mkY.js";
import "./host-env-security-3AOqVC6Z.js";
import "./registry-D5r9Pc2H.js";
import "./manifest-registry-Ix-Y_M6l.js";
import "./dock-DqLW1i1m.js";
import "./message-channel-DOl5pebL.js";
import "./send-C0vRIuJo.js";
import "./plugins-B4oqCXNl.js";
import "./sessions-DTVAB3HG.js";
import "./audio-transcription-runner-9BfH2XEJ.js";
import "./image-CVfYjry_.js";
import "./models-config-CC0tOC2Q.js";
import "./pi-embedded-helpers-BDDoCPNX.js";
import "./sandbox-mimqB061.js";
import "./tool-catalog-Dm7UDWav.js";
import "./chrome-zmDkfOk2.js";
import "./tailscale-C0adBA0J.js";
import "./tailnet-DJJYEayb.js";
import "./ws-BSKgXzsx.js";
import "./auth-CZOQIbmN.js";
import "./credentials-BhjOnp5p.js";
import "./resolve-configured-secret-input-string-D-y9gugr.js";
import "./server-context-BrHTnAHF.js";
import "./frontmatter-NDnrpxIv.js";
import "./env-overrides-lUwusxOI.js";
import "./path-alias-guards-WL7vop6P.js";
import "./skills-D9zp-Tdj.js";
import "./paths-D3yvzAGG.js";
import "./proxy-env-DWmS7QpH.js";
import "./redact-D64ODmQM.js";
import "./errors-BOqrY9lx.js";
import "./fs-safe-XTpVePQ3.js";
import "./image-ops-C7PfPJb_.js";
import "./store-CF7lpR6V.js";
import "./ports-CQ-Y9tpn.js";
import "./trash-CklSLfYF.js";
import "./server-middleware-CiGUwte4.js";
import "./accounts-DCl2uuiL.js";
import "./channel-config-helpers-IZvZlvD6.js";
import "./accounts-BH8lVXqk.js";
import "./send-BAGLdkWp.js";
import "./paths-BXRmsWer.js";
import "./chat-envelope-BkySjpPY.js";
import "./tool-images-L9XyoWe9.js";
import "./tool-display-CIuMJQho.js";
import "./fetch-guard-CEIL7OSV.js";
import "./api-key-rotation-wzo3w_dE.js";
import "./local-roots-B7FLdvEW.js";
import "./model-catalog-Dg5d3V3D.js";
import "./proxy-fetch-B79ly90H.js";
import "./tokens-rNiM9362.js";
import "./deliver-CpeI9Z4f.js";
import "./commands-BYk9iATH.js";
import "./commands-registry-CEWhKPt4.js";
import "./call-Bt8r959b.js";
import "./pairing-token-CYfrO-Yo.js";
import "./send-CLv9RIs5.js";
import "./pi-model-discovery-DYNbVffw.js";
import "./ir-D1mVtfmQ.js";
import "./render-DW1ufaCx.js";
import "./target-errors-Be40pSLb.js";
import "./with-timeout-B8TG8aRr.js";
import "./diagnostic-F6d2aBuS.js";
import "./exec-approvals-allowlist-CzvQC_qV.js";
import "./exec-safe-bin-runtime-policy-B-cncgfZ.js";
import "./exec-approvals-DODENM6Z.js";
import "./nodes-screen-B7ZO3faO.js";
import "./restart-C7ane9OU.js";
import "./system-run-command-BbnKfLEy.js";
import "./runtime-lifecycle-QlGWZQEX.js";
import "./stagger-Bg3n7FlQ.js";
import "./channel-selection-BuAeucPt.js";
import "./plugin-auto-enable-BuUb7WQF.js";
import "./send-DPvyTZ7L.js";
import "./outbound-attachment-BXlVlpCe.js";
import "./fetch-CRD8Jhcq.js";
import "./delivery-queue-Bdmrf3y0.js";
import "./send-DSPzLh7t.js";
import "./pairing-store-C3NC_dEk.js";
import "./session-cost-usage-D8S-fJa3.js";
import "./read-only-account-inspect-BwVlGyMh.js";
import "./sqlite-DC98NVaK.js";
import "./channel-activity-CCCuD5Zr.js";
import "./tables-DpLtvmg5.js";
import "./proxy-CRODgiWq.js";
import "./timeouts-DeXgc5B0.js";
import "./skill-commands-CpM40q42.js";
import "./workspace-dirs-g6Q7CsLa.js";
import "./runtime-config-collectors-CpWVr-ov.js";
import "./command-secret-targets-7ljJh5Io.js";
import "./connection-auth-Ikf5YLLJ.js";
import "./onboard-helpers-qnb8BYWo.js";
import "./prompt-style-D9KjtA09.js";
import "./pairing-labels-C-KU2Np6.js";
import "./memory-cli-1jB6uC2k.js";
import "./manager-DJ_GtFIS.js";
import "./links-BQUKbxMp.js";
import "./cli-utils-B5O9_EBZ.js";
import "./help-format-BWGBPI0G.js";
import "./progress-B8wjC50f.js";
export { countActiveDescendantRuns, countPendingDescendantRuns, countPendingDescendantRunsExcludingRun, isSubagentSessionRunActive, listSubagentRunsForRequester, replaceSubagentRunAfterSteer, resolveRequesterForChildSession, shouldIgnorePostCompletionAnnounceForSession };
