// openclaw@2026.3.8/dist/web-B6HI8eJQ.js
import "./subsystem-kzdGVyce.js";
import "./paths-BfR2LXbA.js";
import "./boolean-BHdNsbzF.js";
import "./auth-profiles-UpqQjKB-.js";
import "./agent-scope-BDXYqF89.js";
import "./utils-B4wShrgI.js";
import "./boundary-file-read-CcBHg_z-.js";
import "./logger-BWbD2ty6.js";
import "./exec-CDUUaRm4.js";
import "./github-copilot-token-CcBrBN3h.js";
import "./host-env-security-blJbxyQo.js";
import "./version-Bxx5bg6l.js";
import "./registry-Gs6xb3DK.js";
import "./manifest-registry-CFnC0yb4.js";
import "./dock-6TSCZqEs.js";
import "./models-config-Ct5bZG7t.js";
import "./tokens-B2BIDASk.js";
import "./compact-D3emcZgv.js";
import "./plugins-v0UiRqAc.js";
import { f as pickWebChannel, l as logWebSelfId, s as WA_WEB_AUTH_DIR, v as webAuthExists } from "./accounts-CFByy36g.js";
import "./channel-config-helpers-Cc37-fbu.js";
import "./send-CbPREZmS.js";
import "./send-Bj6Jq5a4.js";
import "./with-timeout-BbQnczA2.js";
import "./deliver-bv9qu6sz.js";
import "./diagnostic-BjGEEiT0.js";
import "./accounts-Dd357WkV.js";
import "./image-ops-BZX4S9OA.js";
import "./send-oxP8y9QI.js";
import "./pi-model-discovery-CHZFD2LJ.js";
import "./message-channel-FOsBizRL.js";
import "./pi-embedded-helpers-nnm4iOi3.js";
import "./sandbox-D5S9pIUQ.js";
import "./tool-catalog-DAnzECvM.js";
import "./chrome-_dzns-pF.js";
import "./tailscale-DwUR_aBY.js";
import "./tailnet-BEiNr4qb.js";
import "./ws-BzHIlqUk.js";
import "./auth-ztX1XT7i.js";
import "./credentials-UUOBfSK0.js";
import "./resolve-configured-secret-input-string-CALUVGLW.js";
import "./server-context-Cysq8uwv.js";
import "./frontmatter-B5Xx_5Cp.js";
import "./env-overrides-CghEbWu_.js";
import "./path-alias-guards-2B2VwR9Z.js";
import "./skills-BcTP9HTD.js";
import "./paths-Bv0Fd5hm.js";
import "./proxy-env-DBXIh4R2.js";
import "./redact-b9XS0Muh.js";
import "./errors-oTqWODa8.js";
import "./fs-safe-DnLkCsXk.js";
import "./store-Cpd6HHUc.js";
import "./ports-DwmubRiU.js";
import "./trash-CKkKoeyk.js";
import "./server-middleware-D5JzODMR.js";
import "./sessions-Bh025Q5O.js";
import "./paths-D3cXuvcv.js";
import "./chat-envelope-Ba8wEdTK.js";
import "./tool-images-CpDLyo5v.js";
import "./thinking-CHrWXrB4.js";
import "./exec-approvals-allowlist-CmGNghDQ.js";
import "./exec-safe-bin-runtime-policy-BIU2HUmH.js";
import "./model-catalog-RhG2KE3w.js";
import "./fetch-CEXnYKAv.js";
import "./audio-transcription-runner-CkHnvIGC.js";
import "./fetch-guard-W2TcXMQH.js";
import "./image-CiVJHtoW.js";
import "./tool-display-BJIcX_xB.js";
import "./api-key-rotation-BQqyYKRt.js";
import "./proxy-fetch-BJuVlAGN.js";
import "./ir-MPXYeM9z.js";
import "./render-DT-umBSz.js";
import "./target-errors-DkGMVAsH.js";
import "./commands-7gthnTya.js";
import "./commands-registry-Cw8jiBIe.js";
import "./session-cost-usage-0f4Id4RU.js";
import "./session-utils-Cff6Vxd-.js";
import "./sqlite-Q5Gkerf1.js";
import "./call-BfhGytph.js";
import "./pairing-token-Bu2Xt1ri.js";
import "./fetch-Q5enmqDF.js";
import "./pairing-store-v0j3oBoa.js";
import "./exec-approvals-Bc3G-ceB.js";
import "./nodes-screen-DPo01McV.js";
import "./restart-8rBxGr-n.js";
import "./system-run-command-DKHL4lN5.js";
import "./skill-commands-eL6X2qb6.js";
import "./pi-tools.policy-B0vneueD.js";
import "./workspace-dirs-Bi16CKb9.js";
import "./channel-activity-CoKobep6.js";
import "./tables--wgAB_S9.js";
import "./runtime-lifecycle-CI5BW-bB.js";
import "./stagger-BBJI9ikk.js";
import "./channel-selection-CZXcbBEr.js";
import "./plugin-auto-enable-Djxxngde.js";
import "./send-DIPIlK8e.js";
import "./outbound-attachment-DfwwWWQf.js";
import "./delivery-queue-cVcGAG6r.js";
import "./send-DHcDLbo7.js";
import "./proxy-3O29IgJK.js";
import "./timeouts-r52PcxXY.js";
import "./runtime-config-collectors-mVzg9wUg.js";
import "./command-secret-targets-tuZ5jpnm.js";
import "./connection-auth-wdU52rBJ.js";
import "./onboard-helpers-Be9c-Ktq.js";
import "./prompt-style-DMLxb9pN.js";
import "./pairing-labels-D1eHodye.js";
import "./memory-cli-C0jP3BCc.js";
import "./manager-_AfSnopa.js";
import "./links-CVgNb7TN.js";
import "./cli-utils-ChNkua_i.js";
import "./help-format-UvCfCIbo.js";
import "./progress-BVJ8qDnh.js";
import { t as sendMessageWhatsApp } from "./outbound-Deqqmur3.js";
import { n as monitorWebInbox, t as monitorWebChannel } from "./channel-web-Dwj79Wvn.js";
import { i as waitForWaConnection, t as createWaSocket } from "./session--iZwsIQH.js";
import { t as loginWeb } from "./login--ykES9lz.js";
export { WA_WEB_AUTH_DIR, createWaSocket, logWebSelfId, loginWeb, monitorWebChannel, monitorWebInbox, pickWebChannel, sendMessageWhatsApp, waitForWaConnection, webAuthExists };
