// openclaw@2026.3.8/dist/plugin-sdk/subagent-registry-runtime-CWRXxJOO.js
import { a as isSubagentSessionRunActive, c as resolveRequesterForChildSession, i as countPendingDescendantRunsExcludingRun, l as shouldIgnorePostCompletionAnnounceForSession, n as countActiveDescendantRuns, o as listSubagentRunsForRequester, r as countPendingDescendantRuns, s as replaceSubagentRunAfterSteer } from "./dispatch-CM4tRXYq.js";
import "./run-with-concurrency-DKyjYPBF.js";
import "./plugins-BB8tJCZU.js";
import "./accounts-B2dcuvfm.js";
import "./model-auth-BQbQxm_U.js";
import "./logger-D-NoZ0-e.js";
import "./paths-CyBu6eBm.js";
import "./github-copilot-token-CYOZhiYW.js";
import "./thinking-B4hron7F.js";
import "./send-C6DfexIV.js";
import "./send-DHgKdFcP.js";
import "./send-DLISOj2l.js";
import "./image-ops-CyRyYsp0.js";
import "./pi-embedded-helpers-D7aWG1a2.js";
import "./accounts-D9qgbnMC.js";
import "./paths-D_47imm-.js";
import "./deliver-CFjuE9j2.js";
import "./diagnostic-C_yQmSxZ.js";
import "./pi-model-discovery-B6gBxu2i.js";
import "./audio-transcription-runner-D4qZovC_.js";
import "./image-C5phNNN6.js";
import "./chrome-BysAWpFk.js";
import "./skills-lRCwXWwo.js";
import "./path-alias-guards-PaZW0WNH.js";
import "./proxy-env-Ci-s5A4a.js";
import "./redact-DHod2Fyz.js";
import "./errors-BOv7VVr6.js";
import "./fs-safe-Ca2pX9wc.js";
import "./store-CO7bvTgh.js";
import "./tool-images-5h52xKC6.js";
import "./fetch-guard-CatMHzad.js";
import "./api-key-rotation-BDgsszn0.js";
import "./local-roots-Dj0s_gDM.js";
import "./proxy-fetch-DWi0z2To.js";
import "./tokens-C6pUhnym.js";
import "./commands-registry-DsdBUmJR.js";
import "./skill-commands-DHHo44qv.js";
import "./ir-DGHH9wXM.js";
import "./render-hUn-4tdL.js";
import "./target-errors-xEu-KQ_1.js";
import "./send-B3zfH7jf.js";
import "./outbound-attachment-DhsPzcKT.js";
import "./fetch-zYjqExw4.js";
import "./send-B9cgu6Z2.js";
import "./sqlite-BV2UAFoP.js";
import "./channel-activity-BVhRRAlE.js";
import "./tables-C2O_j-f_.js";
import "./proxy-2Si8-2xo.js";
import "./manager-DeHUI--X.js";
export { countActiveDescendantRuns, countPendingDescendantRuns, countPendingDescendantRunsExcludingRun, isSubagentSessionRunActive, listSubagentRunsForRequester, replaceSubagentRunAfterSteer, resolveRequesterForChildSession, shouldIgnorePostCompletionAnnounceForSession };
