// openclaw@2026.3.8/dist/plugin-sdk/subagent-registry-runtime-0Zrs5LOj.js
import { a as isSubagentSessionRunActive, c as resolveRequesterForChildSession, i as countPendingDescendantRunsExcludingRun, l as shouldIgnorePostCompletionAnnounceForSession, n as countActiveDescendantRuns, o as listSubagentRunsForRequester, r as countPendingDescendantRuns, s as replaceSubagentRunAfterSteer } from "./dispatch-DwgTiP0N.js";
import "./send-BlJfRQW2.js";
import "./run-with-concurrency-tODp1hy3.js";
import "./plugins-e62olPre.js";
import "./model-auth-BQ4Cp3j8.js";
import "./logger-D-NoZ0-e.js";
import "./paths-CyBu6eBm.js";
import "./github-copilot-token-CYOZhiYW.js";
import "./thinking-D-7Pwj4y.js";
import "./accounts-Dp-gY4hm.js";
import "./ssrf-BOgvekn1.js";
import "./fetch-guard-DqcetWYz.js";
import "./send-BunVoKbK.js";
import "./send-qRJIpBTo.js";
import "./image-ops-BrTKipfG.js";
import "./pi-embedded-helpers-C7GCMUPD.js";
import "./accounts-DLGN1oUP.js";
import "./paths-Cn4jOpOv.js";
import "./deliver-DXo2B1Mw.js";
import "./diagnostic-CRQnfytD.js";
import "./pi-model-discovery-DUEz1g-Y.js";
import "./audio-transcription-runner-DsiU1gkv.js";
import "./image-Cj2Jg3ch.js";
import "./chrome-D5Lr0QX7.js";
import "./skills-B_DPlC7t.js";
import "./path-alias-guards-8uK4lP7_.js";
import "./redact-B1C1rXrR.js";
import "./errors-CReGQ7X8.js";
import "./fs-safe-CfCxUVsz.js";
import "./store-C5u6XrXo.js";
import "./tool-images-B2l1KmaR.js";
import "./api-key-rotation-CnwWrazx.js";
import "./local-roots-DR4GvGT2.js";
import "./proxy-fetch-DWi0z2To.js";
import "./tokens-C6pUhnym.js";
import "./commands-registry-CSnzvOBa.js";
import "./skill-commands-CECuzcwF.js";
import "./ir-BzuJwcEb.js";
import "./render-hUn-4tdL.js";
import "./target-errors-SxlmZyYv.js";
import "./send-BxC7n_GZ.js";
import "./outbound-attachment-Pxc0iRtl.js";
import "./fetch-CaJ6Ir7Q.js";
import "./send-ChRjwaf8.js";
import "./sqlite-BW3YgB8E.js";
import "./channel-activity-eVzrH2pR.js";
import "./tables-BFWhW4Ke.js";
import "./proxy-2Si8-2xo.js";
import "./manager-Bf96hOQw.js";
export { countActiveDescendantRuns, countPendingDescendantRuns, countPendingDescendantRunsExcludingRun, isSubagentSessionRunActive, listSubagentRunsForRequester, replaceSubagentRunAfterSteer, resolveRequesterForChildSession, shouldIgnorePostCompletionAnnounceForSession };
