// openclaw@2026.3.8/dist/plugin-sdk/web-CJp2fj9G.js
import "./run-with-concurrency-CvBOjUE3.js";
import "./config-upduwSBz.js";
import "./logger-CyvuRp4f.js";
import "./paths-D6tDENa_.js";
import { f as pickWebChannel, l as logWebSelfId, s as WA_WEB_AUTH_DIR, v as webAuthExists } from "./accounts-CKWobc8T.js";
import "./plugins-f1v1_i9D.js";
import "./send-GWvi69z3.js";
import "./common-BsU_fpVA.js";
import "./send-eprYBFVV.js";
import "./thinking-B4aBdZa6.js";
import "./image-ops-CSd6ioRA.js";
import "./pi-embedded-helpers-C-tB5b5B.js";
import "./accounts-CYGUsYvM.js";
import "./reply-UQ7w3uFC.js";
import "./github-copilot-token-xlpfBCoP.js";
import "./paths-0c91o_Js.js";
import "./send-DnRAcaR9.js";
import { n as monitorWebInbox, t as monitorWebChannel } from "./channel-web-D2YXVE7V.js";
import "./tokens-DwHJ1omO.js";
import "./audio-transcription-runner-BtnWjA5B.js";
import "./image-BBszd7R_.js";
import "./chrome-CN1wAraE.js";
import "./skills-DrcY1qHs.js";
import "./path-alias-guards-RSyiESJ8.js";
import "./proxy-env-BR0_ihdh.js";
import "./redact-66Hhagy9.js";
import "./errors-C6T03XVS.js";
import "./fs-safe-K-umaOqX.js";
import "./store-XqCQC_cj.js";
import "./tool-images-BswvL9EH.js";
import "./fetch-guard-28z9IJbo.js";
import "./api-key-rotation-CQzQxcZg.js";
import "./local-roots-BxJbYi6l.js";
import "./proxy-fetch-BPy37MWG.js";
import "./deliver-CkModk8N.js";
import "./commands-registry-DNDsmNbw.js";
import "./skill-commands-RBudNjnM.js";
import "./pi-model-discovery-Bd-m77Xt.js";
import "./ir-DjYCWMG5.js";
import "./render-7C7EDC8_.js";
import "./diagnostic-DL-FXUFm.js";
import "./send-BfzX5ud_.js";
import "./outbound-attachment-nDi105-h.js";
import "./fetch-Bo3QUo20.js";
import "./send-BBoB1UfA.js";
import "./sqlite-BlKQcZG1.js";
import "./channel-activity-DuNlXOPl.js";
import "./tables-CZT3J_Ar.js";
import "./proxy-BzwL4n0W.js";
import "./manager-BuLFYJ4s.js";
import { t as sendMessageWhatsApp } from "./outbound-PXQWn1s7.js";
import { i as waitForWaConnection, t as createWaSocket } from "./session-NXCwEOlL.js";
import { t as loginWeb } from "./login-DFEL9wW8.js";
export { WA_WEB_AUTH_DIR, createWaSocket, logWebSelfId, loginWeb, monitorWebChannel, monitorWebInbox, pickWebChannel, sendMessageWhatsApp, waitForWaConnection, webAuthExists };
